File: apps/agent/src/tools/updateUserProfile.ts
import { FunctionTool } from '@google/adk';
import { z } from 'zod';
import type { UserProfileStore } from '../storage/userProfile.js';

/**
 * Schema-free profile tool. The coach can invent any dotted path
 * (`pets.name`, `morning_routine.coffee_first`, `volunteering`, ...) and
 * persist it to the user's YAML. See memory/feedback_yaml_schema_free.md.
 *
 * `value` is still a nullable string for the LLM call-site. `resolveValue`
 * does two lightweight coercions as a convenience:
 *   - `age` → number (common enough to bother)
 *   - `goals.{short,medium,long}_term` → JSON-parsed array of strings
 *     (the coach wants lists for goal tiers; the user's schema-free
 *     stance doesn't preclude convenience heuristics)
 * Everything else passes through as a string.
 */
export function createUpdateUserProfileTool(deps: {
  store: UserProfileStore;
  uid: string;
}): FunctionTool {
  const parameters = z.object({
    path: z
      .string()
      .min(1)
      .describe(
        'Dotted path into the user profile YAML (e.g. "name", "family.children", "pets.species", "volunteering"). Invent new keys freely when a fact doesn\'t fit an existing slot — the profile has no fixed schema.',
      ),
    value: z
      .string()
      .nullable()
      .describe(
        'New value as a string. Numbers: stringified (age is coerced back to number). Goals tiers (goals.short_term etc): JSON array string. Null clears the field.',
      ),
  });

  // biome-ignore lint/suspicious/noExplicitAny: zod instance nominal mismatch with ADK generics
  return new FunctionTool<any>({
    name: 'update_user_profile',
    description:
      'Persist a structured fact about the user (or someone in their life) to their ' +
      'profile YAML. Call this PROACTIVELY the FIRST time you hear any of: ' +
      '(a) a name of someone in their life — partner, child, sibling, parent, close ' +
      'friend, regular colleague (e.g. "Wren\'s parents\' evening" → save Wren); ' +
      "(b) the user's own identity facts — name, city/postcode, occupation, key " +
      'health context (injuries, conditions); ' +
      "(c) interests, hobbies, regular sports — theirs OR a family member's; " +
      '(d) routines ("Tuesday yoga", "Sunday long runs"); ' +
      '(e) strong preferences/dislikes ("I never drink coffee after 2pm"). ' +
      'Use any sensible dotted path; invent freely (family.children[0].name, ' +
      'family.partner.name, occupation.title, health.injuries.calf_strain, ' +
      'preferences.coffee_cutoff). The profile has no fixed schema. ' +
      'Heuristic: if a fact would still matter next month, capture it now. ' +
      'Never announce ("let me note that") — save silently and continue talking.',
    parameters,
    execute: async (input: unknown) => {
      const { path, value } = input as { path: string; value: string | null };
      try {
        const resolved = resolveValue(path, value);
        const profile = await deps.store.updatePath(deps.uid, path, resolved);
        return {
          status: 'ok' as const,
          updated_path: path,
          new_value: resolved,
          profile_after: profile,
        };
      } catch (err) {
        return {
          status: 'error' as const,
          message: err instanceof Error ? err.message : String(err),
        };
      }
    },
  });
}

const GOAL_LIST_PATHS = new Set(['goals.short_term', 'goals.medium_term', 'goals.long_term']);

function resolveValue(path: string, value: string | null): unknown {
  if (value === null) return null;
  if (path === 'age') {
    const n = Number(value);
    if (Number.isNaN(n)) throw new Error(`age must be numeric, got "${value}"`);
    return n;
  }
  if (GOAL_LIST_PATHS.has(path)) {
    try {
      const parsed = JSON.parse(value);
      if (!Array.isArray(parsed) || !parsed.every((x) => typeof x === 'string')) {
        throw new Error('must be a JSON array of strings');
      }
      return parsed;
    } catch (err) {
      throw new Error(
        `goals.*_term must be a JSON array of strings, got ${JSON.stringify(value)}: ${err instanceof Error ? err.message : String(err)}`,
      );
    }
  }
  return value;
}

File: apps/agent_py/src/lifecoach_agent/tools/update_user_profile.py
"""update_user_profile tool — schema-free profile writes.

The coach can invent any dotted path (`pets.name`,
`morning_routine.coffee_first`, `volunteering`, …) and persist it to
the user's YAML. See memory/feedback_yaml_schema_free.md.

`value` is still a nullable string at the LLM call-site; `_resolve_value`
applies two convenience coercions:
  - `age` → number
  - `goals.{short,medium,long}_term` → JSON-parsed list of strings
Everything else passes through as a string.

Mirrors `apps/agent/src/tools/updateUserProfile.ts`.
"""

from __future__ import annotations

import json
import logging
from datetime import UTC, datetime
from typing import Any

from lifecoach_agent.storage.profile_history import ProfileHistoryEntry, ProfileHistoryStore
from lifecoach_agent.storage.user_profile import UserProfileStore

_log = logging.getLogger(__name__)


_GOAL_LIST_PATHS = frozenset({"goals.short_term", "goals.medium_term", "goals.long_term"})


def _resolve_value(path: str, value: str | None) -> Any:
    if value is None:
        return None
    if path == "age":
        try:
            return int(value)
        except ValueError:
            try:
                return float(value)
            except ValueError as e:
                raise ValueError(f'age must be numeric, got "{value}"') from e
    if path in _GOAL_LIST_PATHS:
        try:
            parsed = json.loads(value)
        except json.JSONDecodeError as err:
            raise ValueError(
                f"goals.*_term must be a JSON array of strings, got {value!r}: {err}"
            ) from err
        if not isinstance(parsed, list) or not all(isinstance(x, str) for x in parsed):
            raise ValueError(f"goals.*_term must be a JSON array of strings, got {value!r}")
        return parsed
    return value


def create_update_user_profile_tool(
    *,
    store: UserProfileStore,
    uid: str,
    history: ProfileHistoryStore | None = None,
) -> Any:
    """Build a closure-bound `update_user_profile` tool. Optional
    `history` audit-log store; production wiring always supplies it."""

    async def update_user_profile(path: str, value: str | None) -> dict[str, Any]:
        """Persist a structured fact about the user (or someone in their
        life) to their profile YAML. Call this PROACTIVELY the FIRST
        time you hear any of:
        (a) a name of someone in their life — partner, child, sibling,
        parent, close friend, regular colleague (e.g. "Maya's parents'
        evening" → save Maya);
        (b) the user's own identity facts — name, city/postcode,
        occupation, key health context (injuries, conditions);
        (c) interests, hobbies, regular sports — theirs OR a family
        member's;
        (d) routines ("Tuesday yoga", "Sunday long runs");
        (e) strong preferences/dislikes ("I never drink coffee after 2pm").
        Use any sensible dotted path; invent freely
        (family.children[0].name, family.partner.name, occupation.title,
        health.injuries.calf_strain, preferences.coffee_cutoff). The
        profile has no fixed schema. Heuristic: if a fact would still
        matter next month, capture it now. Never announce ("let me note
        that") — save silently and continue talking.

        Args:
            path: Dotted path into the user profile YAML (e.g. "name",
                "family.children", "pets.species", "volunteering").
                Invent new keys freely when a fact doesn't fit an
                existing slot — the profile has no fixed schema.
            value: New value as a string. Numbers: stringified (age is
                coerced back to number). Goals tiers
                (goals.short_term etc): JSON array string. Null clears
                the field.
        """
        try:
            resolved = _resolve_value(path, value)
            before = await store.read_path(uid, path)
            profile = await store.update_path(uid, path, resolved)
            at = datetime.now(UTC).isoformat().replace("+00:00", "Z")
            if history is not None:
                try:
                    await history.append(
                        uid,
                        ProfileHistoryEntry(
                            path=path,
                            before=before,
                            after=resolved,
                            at=at,
                        ),
                    )
                except Exception as log_err:  # noqa: BLE001
                    _log.error("profile.history.append failed: %s", log_err)
            return {
                "status": "ok",
                "updated_path": path,
                "new_value": resolved,
                "previous_value": before,
                "modified_at": at,
                "profile_after": profile,
            }
        except Exception as err:  # noqa: BLE001
            return {"status": "error", "message": str(err)}

    from google.adk.tools import FunctionTool

    return FunctionTool(update_user_profile)
